File: internal/maintenance/health.go
package maintenance

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
)

// daemonTailBytes caps how much of the tail of daemon.log is read — the last N
// lines come from this window, never the whole (potentially huge) file.
const (
	daemonTailBytes = 64 << 10
	daemonTailLines = 40
)

// knownFlagFiles is the allowlist of mode-flag dotfiles the health panel shows —
// never a raw enumeration of every dotfile (avoids surfacing stray names).
var knownFlagFiles = []string{".caveman-active", ".ponytail-active"}

// FlagFile is one known mode-flag dotfile's presence + content.
type FlagFile struct {
	Name    string `json:"name"`
	Present bool   `json:"present"`
	Content string `json:"content"`
}

// HealthStatus is the read-only health snapshot for the W14 panel. Times are
// file mtimes (ms) — a reliable signal regardless of file content format.
type HealthStatus struct {
	LastCleanupMs      float64    `json:"lastCleanupMs"` // .last-cleanup mtime, 0 if absent
	LastCleanupRaw     string     `json:"lastCleanupRaw"`
	LastUpdateRaw      string     `json:"lastUpdateRaw"`
	LastUpdateStatus   string     `json:"lastUpdateStatus"`
	LastUpdateVersion  string     `json:"lastUpdateVersion"`
	LastUpdateParseErr bool       `json:"lastUpdateParseErr"`
	DaemonPresent      bool       `json:"daemonPresent"`
	DaemonLastWriteMs  float64    `json:"daemonLastWriteMs"`
	DaemonTail         []string   `json:"daemonTail"`
	Flags              []FlagFile `json:"flags"`
	// SchedulerInterval + LastAutoCleanupMs are the W32 in-app scheduler status,
	// populated by the service layer from config (this pure reader leaves them
	// zero-valued). SchedulerInterval is "off"|"weekly"|"monthly";
	// LastAutoCleanupMs is the app's OWN last policy-clean run (0 = never).
	SchedulerInterval string  `json:"schedulerInterval"`
	LastAutoCleanupMs float64 `json:"lastAutoCleanupMs"`
}

// MaintenanceHealth reads the four health surfaces read-only (no side effects).
// Missing files yield zero/absent states, never errors.
func MaintenanceHealth(root string) (HealthStatus, error) {
	h := HealthStatus{Flags: []FlagFile{}, DaemonTail: []string{}}

	if info, err := os.Lstat(filepath.Join(root, ".last-cleanup")); err == nil && !info.IsDir() {
		h.LastCleanupMs = float64(info.ModTime().UnixMilli())
		if data, err := os.ReadFile(filepath.Join(root, ".last-cleanup")); err == nil {
			h.LastCleanupRaw = strings.TrimSpace(string(data))
		}
	}

	if data, err := os.ReadFile(filepath.Join(root, ".last-update-result.json")); err == nil {
		h.LastUpdateRaw = strings.TrimSpace(string(data))
		var parsed struct {
			Status  string `json:"status"`
			Version string `json:"version"`
		}
		if json.Unmarshal(data, &parsed) != nil {
			h.LastUpdateParseErr = true
		} else {
			h.LastUpdateStatus = parsed.Status
			h.LastUpdateVersion = parsed.Version
		}
	}

	daemonPath := filepath.Join(root, "daemon.log")
	if info, err := os.Lstat(daemonPath); err == nil && !info.IsDir() && info.Mode()&os.ModeSymlink == 0 {
		h.DaemonPresent = true
		h.DaemonLastWriteMs = float64(info.ModTime().UnixMilli())
		h.DaemonTail = tailLines(daemonPath, info.Size())
	}

	for _, name := range knownFlagFiles {
		flag := FlagFile{Name: name}
		if data, err := os.ReadFile(filepath.Join(root, name)); err == nil {
			flag.Present = true
			flag.Content = strings.TrimSpace(string(data))
		}
		h.Flags = append(h.Flags, flag)
	}

	return h, nil
}

// tailLines returns the last daemonTailLines lines of path, reading only the
// final daemonTailBytes window (never the whole file).
func tailLines(path string, size int64) []string {
	f, err := os.Open(path)
	if err != nil {
		return []string{}
	}
	defer f.Close()

	readLen := int64(daemonTailBytes)
	offset := int64(0)
	if size > readLen {
		offset = size - readLen
	} else {
		readLen = size
	}
	buf := make([]byte, readLen)
	if _, err := f.ReadAt(buf, offset); err != nil && err.Error() != "EOF" {
		return []string{}
	}

	lines := strings.Split(strings.TrimRight(string(buf), "\n"), "\n")
	if offset > 0 && len(lines) > 0 {
		lines = lines[1:] // drop the partial first line from mid-file
	}
	if len(lines) > daemonTailLines {
		lines = lines[len(lines)-daemonTailLines:]
	}
	return lines
}

File: src-tauri/src/maintenance/health.rs
//! Ports `internal/maintenance/health.go` — the READ-ONLY health snapshot for
//! the maintenance panel. No side effects: missing files yield zero/absent
//! states, never errors. Reads `.last-cleanup`, `.last-update-result.json`, the
//! tail of `daemon.log`, and the known mode-flag dotfiles (an allowlist, never a
//! raw dotfile enumeration). Guards reproduced verbatim (invariant #3).

use std::fs::{self, File, Metadata};
use std::os::unix::fs::FileExt;
use std::path::Path;
use std::time::UNIX_EPOCH;

use serde::{Deserialize, Serialize};

/// Caps how much of the tail of daemon.log is read — the last N lines come from
/// this window, never the whole file. Mirrors `daemonTailBytes`/`daemonTailLines`.
const DAEMON_TAIL_BYTES: u64 = 64 << 10;
const DAEMON_TAIL_LINES: usize = 40;

/// Allowlist of mode-flag dotfiles the health panel shows. Mirrors
/// `knownFlagFiles`.
const KNOWN_FLAG_FILES: [&str; 2] = [".caveman-active", ".ponytail-active"];

/// One known mode-flag dotfile's presence + content. Mirrors `FlagFile`.
#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FlagFile {
    pub name: String,
    pub present: bool,
    pub content: String,
}

/// The read-only health snapshot. Times are file mtimes (ms). Mirrors
/// `HealthStatus`. `scheduler_interval` + `last_auto_cleanup_ms` are the in-app
/// scheduler status the SERVICE layer populates from config — this pure reader
/// leaves them zero-valued (settable, but never set here).
#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HealthStatus {
    pub last_cleanup_ms: f64,
    pub last_cleanup_raw: String,
    pub last_update_raw: String,
    pub last_update_status: String,
    pub last_update_version: String,
    pub last_update_parse_err: bool,
    pub daemon_present: bool,
    pub daemon_last_write_ms: f64,
    pub daemon_tail: Vec<String>,
    pub flags: Vec<FlagFile>,
    pub scheduler_interval: String,
    pub last_auto_cleanup_ms: f64,
}

/// Reads the four health surfaces read-only (no side effects). Missing files
/// yield zero/absent states, never errors. Mirrors `MaintenanceHealth`.
pub fn maintenance_health(root: &str) -> Result<HealthStatus, String> {
    let mut h = HealthStatus::default();
    let root = Path::new(root);

    // .last-cleanup: mtime + trimmed raw content.
    let last_cleanup = root.join(".last-cleanup");
    if let Ok(meta) = fs::symlink_metadata(&last_cleanup) {
        if !meta.is_dir() {
            h.last_cleanup_ms = mtime_ms(&meta);
            if let Ok(data) = fs::read(&last_cleanup) {
                h.last_cleanup_raw = trimmed(&data);
            }
        }
    }

    // .last-update-result.json: raw + parsed {status, version}; parse failure is
    // a flag, not fatal.
    if let Ok(data) = fs::read(root.join(".last-update-result.json")) {
        h.last_update_raw = trimmed(&data);
        #[derive(Deserialize)]
        struct Parsed {
            #[serde(default)]
            status: String,
            #[serde(default)]
            version: String,
        }
        match serde_json::from_slice::<Parsed>(&data) {
            Ok(p) => {
                h.last_update_status = p.status;
                h.last_update_version = p.version;
            }
            Err(_) => h.last_update_parse_err = true,
        }
    }

    // daemon.log: present + mtime + last-N-lines tail. A symlink is refused.
    let daemon_path = root.join("daemon.log");
    if let Ok(meta) = fs::symlink_metadata(&daemon_path) {
        if !meta.is_dir() && !meta.file_type().is_symlink() {
            h.daemon_present = true;
            h.daemon_last_write_ms = mtime_ms(&meta);
            h.daemon_tail = tail_lines(&daemon_path, meta.len());
        }
    }

    // Known flag dotfiles: always listed (present=false when absent).
    for name in KNOWN_FLAG_FILES {
        let mut flag = FlagFile {
            name: name.to_string(),
            present: false,
            content: String::new(),
        };
        if let Ok(data) = fs::read(root.join(name)) {
            flag.present = true;
            flag.content = trimmed(&data);
        }
        h.flags.push(flag);
    }

    Ok(h)
}

/// Returns the last `DAEMON_TAIL_LINES` lines of path, reading only the final
/// `DAEMON_TAIL_BYTES` window. Mirrors `tailLines`.
fn tail_lines(path: &Path, size: u64) -> Vec<String> {
    let Ok(f) = File::open(path) else {
        return Vec::new();
    };

    let mut read_len = DAEMON_TAIL_BYTES;
    let mut offset = 0u64;
    if size > read_len {
        offset = size - read_len;
    } else {
        read_len = size;
    }

    let mut buf = vec![0u8; read_len as usize];
    // Sized to exactly the available bytes, so a full read is expected; any
    // error (mirroring Go's non-EOF bail) yields an empty tail.
    if f.read_exact_at(&mut buf, offset).is_err() {
        return Vec::new();
    }

    let text = String::from_utf8_lossy(&buf);
    let trimmed_text = text.trim_end_matches('\n');
    let mut lines: Vec<String> = trimmed_text.split('\n').map(|s| s.to_string()).collect();
    if offset > 0 && !lines.is_empty() {
        lines.remove(0); // drop the partial first line from mid-file
    }
    if lines.len() > DAEMON_TAIL_LINES {
        lines = lines[lines.len() - DAEMON_TAIL_LINES..].to_vec();
    }
    lines
}

/// File mtime in epoch-ms as a float. Mirrors `info.ModTime().UnixMilli()`.
fn mtime_ms(meta: &Metadata) -> f64 {
    meta.modified()
        .ok()
        .and_then(|t| t.duration_since(UNIX_EPOCH).ok())
        .map(|d| d.as_millis() as f64)
        .unwrap_or(0.0)
}

/// Trims surrounding ASCII whitespace, matching Go's `strings.TrimSpace` for the
/// common cases (space/tab/newline/CR).
fn trimmed(data: &[u8]) -> String {
    String::from_utf8_lossy(data).trim().to_string()
}

#[cfg(test)]
#[path = "health_tests.rs"]
mod health_tests;
